extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+        new DimensionsSpec(dimensionsAndAggregators.lhs),
+        dimensionsAndAggregators.rhs.toArray(new AggregatorFactory[0]),
+        makeGranularitySpecForIngestion(isRollupQ, querySpec.getQuery()),
+        new TransformSpec(null, Collections.emptyList())
+    );
+  }
+
+  private static GranularitySpec makeGranularitySpecForIngestion(boolean isRollupQ, Query<?> query)
+  {
+    if (!isRollupQ) {
+      return new ArbitraryGranularitySpec(Granularities.NONE, false, Intervals.ONLY_ETERNITY);
+    } else {
+      final String queryGranularity = query.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD_GRANULARITY, "");
+
+      if (checkIfTimeColumnsAreEqual((GroupByQuery) query) && !queryGranularity.isEmpty()) {
+        return new ArbitraryGranularitySpec(
+            Granularity.fromString(queryGranularity),
+            true,
+            Intervals.ONLY_ETERNITY
+        );
+      }
+      return new ArbitraryGranularitySpec(Granularities.NONE, true, Intervals.ONLY_ETERNITY);
+    }
+  }
+
+
+  /**
+   * Checks if the time columns present in the groupByQuery context are same. One is set by
+   * {@link DruidQuery#toGroupByQuery()} and the other is set by
+   * {@link MSQTaskQueryMaker#runQuery(DruidQuery)}
+   *
+   * @return true if both groupByQuery context values are present and equal else returns false.
+   */
+  private static boolean checkIfTimeColumnsAreEqual(GroupByQuery groupByQuery)
+  {
+    final String msqTimeColumn = groupByQuery.getContextValue(QueryKitUtils.CTX_TIME_COLUMN_NAME, "");
+    if (msqTimeColumn.isEmpty()) {
+      return false;
+    }
+    return msqTimeColumn.equals(groupByQuery.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD));
+  }
+
+  /**
+   * Checks if segments generated by the insert query can be rolled up futher.
+   *
+   * @param query
+   *
+   * @return
+   */
+  private static boolean isRollupQuery(Query<?> query)
+  {
+    return MultiStageQueryContext.isFinalizeAggregations(query.getQueryContext()) == false
+           && query.getContextBoolean(GroupByQueryConfig.CTX_KEY_ENABLE_MULTI_VALUE_UNNESTING, true) == false
+           && query instanceof GroupByQuery;
+  }
+
+  private static boolean isInlineResults(final MSQSpec querySpec)
+  {
+    return querySpec.getDestination() instanceof TaskReportMSQDestination;
+  }
+
+  private static boolean isTimeBucketedIngestion(final MSQSpec querySpec)
+  {
+    return MSQControllerTask.isIngestion(querySpec)
+           && !((DataSourceMSQDestination) querySpec.getDestination()).getSegmentGranularity()
+                                                                      .equals(Granularities.ALL);
+  }
+
+  /**
+   * Compute shard columns for {@link DimensionRangeShardSpec}. Returns an empty list if range-based sharding
+   * is not applicable.
+   */
+  private static List<String> computeShardColumns(
+      final RowSignature signature,
+      final ClusterBy clusterBy,
+      final ColumnMappings columnMappings
+  )
```

Comment:
So, I'm reading now about creating the query "plan". In the same class, I read about query execution and results resporting, as well as event handling.

I wonder, does Overlord require that all logic for a task be in a single class? Probably not? That suggestion above about dividing up the logic would help us poor reviews avoid the need to keep so many factors in mind in a single class.

Reply:
> I wonder, does Overlord require that all logic for a task be in a single class?

No, it doesn't, of course. It's already split up to some degree: all of the querykit and kernel packages are bits of logic ultimately used by this ControllerImpl. But I agree it'd be good to split up this class further. My preference would be to do it in follow-ups. Probably the first thing to tackle would be to isolate all the ingestion-related planning into its own class (i.e. what locks to acquire upfront, whether or not the ingestion is rollup-based, what the segment schema is going to be, etc). That would simplify the main class quite a bit.